### ISSUE TITLE
[Task]: POC for using proxies to store data tree values

### ISSUE THREAD

**Issue by nidhi-nair** [appsmithorg/appsmith#11095]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

We're currently storing each entity in the app on the data tree and simply use the dependency map to figure out which entities might have been updated during an evaluation. This does not account for conditional paths. If we use proxies instead, we'll be able to customize the setter to trigger re-evaluation of dependents for only those entities that have actually been set in an eval.

Being able to use proxies will also help us to possibly allow user-defined events as a feature.

**Comment by Rishabh-Rathod** [comment]
> simply use the dependency map to figure out which entities might have been updated during an evaluation.

We use a dependency map to know which all property needs re-evaluation after an evaluation of their dependent property. 

For eg:- 
If `Input.defaultText` = `{{Api.data.users[0].name}}` then the dependency map would be 
```json
{
  "Input.defaultText" : ["Api.data"]
}
```
Every time `Api.data` is evaluated, we also re-evaluate `Input.defaultText` without the dependency map we won't be knowing which property to re-evaluate and hence other values would remain stale. 

>  If we use proxies instead, we'll be able to customize the setter to trigger re-evaluation of dependents for only those entities that have actually been set in an eval.

Current architecture also re-evaluates the dependents. 

> Being able to use proxies will also help us to possibly allow user-defined events as a feature.

@nidhi-nair could you please elaborate more on `user-defined events`?
Also, could you please confirm if the task is valid anymore or not?

**Comment by Rishabh-Rathod** [comment]
After trying out the javascript proxy for the JSObject mutation feature, we realized that proxy has its own caveats. 

- A proxy object cannot be structureCloned.
- Worker needs to send the data back to the main thread using post message where structureCloning happens and it fails to send the data.
- To avoid this issue, we need to write custom logic to remove the proxy object from the data. 
- As it was nested logic and there was no way to know if the data being sent had a proxy object without looping through it, this became an expensive task.

Hence, later on, we decided to remove proxy implementation in favor of getters and setters.
